### ISSUE TITLE
DistributedCache<TCacheItem> should be an adapter instead of inheriting from DistributedCache<TCacheItem, string>

### ISSUE THREAD

**Issue by hikalkan** [abpframework/abp#13322]
Otherwise, we will have two different instances when we inject `IDistributedCache<T>` and `IDistributedCache<T, string>` (`string` is the cache key type here). See the registration:

![image](https://user-images.githubusercontent.com/1210527/179214286-607b2083-2f07-41bc-ba00-1e0eb6e2d468.png)

Current `DistributedCache<TCacheItem>` inherits from `DistributedCache<TCacheItem, string>`:

![image](https://user-images.githubusercontent.com/1210527/179214253-bcd4ac79-beaf-4265-a1cc-2fb45d2cce42.png)

It should wrap `DistributedCache<TCacheItem, string>` instead, like:

![image](https://user-images.githubusercontent.com/1210527/179214170-7055a15b-8bcb-4d5a-a30f-a44a8e90da06.png)

So, basically, we will re-implement `IDistributedCache<TCacheItem>` and redirect all calls to the injected `DistributedCache<TCacheItem, string>` instance.

Actually, this is not a problem if we use redis cache, or we always use `DistributedCache` in a consistent way. But if we use in-memory cache (which is default) and use different interfaces for the same cache item type, we will have two cache instances, which leads to inconsistent data.

Please add test to prove my theory first, then fix it.

**Comment by maliming** [comment]
hi

I checked. The `IDistributedCache(memory,redis,sql)` is a `singleton`, so there is no problem.

**Comment by hikalkan** [comment]
I decided to implement this. Because, I think to add in-memory cache layer (using a dictionary object) to distributed cache and it requires that we should access the same singleton instance.

**Comment by hikalkan** [comment]
done with https://github.com/abpframework/abp/pull/13845
